src/ocf_composite_volume.c

```diff
@@ -0,0 +1,357 @@
+{
+	ocf_volume_t cvolume = ocf_io_get_volume(io);
+	struct ocf_composite_volume *composite = ocf_volume_get_priv(cvolume);
+	struct ocf_composite_volume_io *cio = ocf_io_get_priv(io);
+	uint64_t begin, io_offset;
+	int i;
+
+	cio->data = data;
+
+	for (i = cio->begin_member; i < cio->end_member; i++) {
+		begin = i > 0 ? composite->lookup[i - 1] : 0;
+		io_offset = (io->addr > begin ? 0 : begin - io->addr) + offset;
+		ocf_io_set_data(cio->io[i], data, io_offset);
+	}
+
+	return 0;
+}
+
+static ctx_data_t *ocf_composite_io_get_data(struct ocf_io *io)
+{
+	struct ocf_composite_volume_io *cio = ocf_io_get_priv(io);
+
+	return cio->data;
+}
+
+const struct ocf_volume_properties ocf_composite_volume_properties = {
+	.name = "OCF Composite",
+	.io_priv_size = sizeof(struct ocf_composite_volume_io),
+	.volume_priv_size = sizeof(struct ocf_composite_volume),
+	.caps = {
+		.atomic_writes = 0,
+	},
+	.ops = {
+		.submit_io = ocf_composite_volume_submit_io,
+		.submit_flush = ocf_composite_volume_submit_flush,
+		.submit_discard = ocf_composite_volume_submit_discard,
+		.submit_metadata = NULL,
+
+		.open = ocf_composite_volume_open,
+		.close = ocf_composite_volume_close,
+		.get_max_io_size = ocf_composite_volume_get_max_io_size,
+		.get_length = ocf_composite_volume_get_byte_length,
+	},
+	.io_ops = {
+		.set_data = ocf_composite_io_set_data,
+		.get_data = ocf_composite_io_get_data,
+	},
+	.deinit = NULL,
+};
+
+static int ocf_composite_io_allocator_init(ocf_io_allocator_t allocator,
+		uint32_t priv_size, const char *name)
+{
+	return ocf_io_allocator_default_init(allocator, priv_size, name);
+}
+
+static void ocf_composite_io_allocator_deinit(ocf_io_allocator_t allocator)
+{
+	ocf_io_allocator_default_deinit(allocator);
+}
+
+static void *ocf_composite_io_allocator_new(ocf_io_allocator_t allocator,
+		ocf_volume_t cvolume, ocf_queue_t queue,
+		uint64_t addr, uint32_t bytes, uint32_t dir)
+{
+	struct ocf_composite_volume *composite = ocf_volume_get_priv(cvolume);
+	struct ocf_composite_volume_io *cio;
+	struct ocf_io_internal *ioi;
+	uint64_t begin, end;
+	uint64_t io_addr;
+	uint32_t io_bytes;
+	int i;
+
+	ioi = ocf_io_allocator_default_new(allocator, cvolume, queue,
+			addr, bytes, dir);
+	if (!ioi)
+		return NULL;
+
+	cio = ocf_io_get_priv(&ioi->io);
+
+	if (bytes == 0) {
+		/* Flush io - allocate io for each volume */
+		for (i = 0; i < composite->members_cnt; i++) {
+			cio->io[i] = ocf_io_new(&composite->member[i].volume,
+					queue, 0, 0, dir, 0, 0);
+			if (!cio->io[i])
+				goto err;
+		}
+		cio->begin_member = 0;
+		cio->end_member = composite->members_cnt;
+
+		return ioi;
+	}
+
+	for (i = 0; i < composite->members_cnt; i++) {
+		if (addr < composite->lookup[i]) {
+			cio->begin_member = i;
+			break;
+		}
+	}
+
+	for (; i < composite->members_cnt; i++) {
+		begin = i > 0 ? composite->lookup[i - 1] : 0;
+		end = composite->lookup[i];
+
+		io_addr = addr > begin ? addr - begin : 0;
+		io_bytes = OCF_MIN(bytes, end - addr);
```

Comment:
I think your way of calculating `io_bytes` may be incorrect. Consider the following example with composite volume which consists of 4 volumes of sizes 5, 8, 13 and 4 and an IO with addr 11 and size 16 (the blue color on the diagram). Only the first IO has an appropriate range. `io_bytes` for the second and third ones is invalid. 
![image](https://user-images.githubusercontent.com/25607940/167397750-ec49e707-67bf-4b6d-b744-fa831b427aec.png)

Reply:
@mmichal10 Regarding the graphical representation of I/O ranges at the top of your picture. The second volume is depicted as 9 fields wide instead of 8. The I/O range in blue starts at byte 12 instead of 11.

Reply:
Other than that, I agree with Michał. IO size is is miscalculated.

Reply:
I think this logic would be easier to follow if you had variables for current address and remaining size (incrementing/decrementing addr and bytes would work):

```
io_addr = addr
io_bytes = OCF_MIN(bytes, end - addr);
...
addr += io_bytes
bytes -= io_bytes
```